File: Tests/Review1/dijkstra/dijkstra.ts
type EdgeMap = Record<string, number>;
type Graph = Record<string, EdgeMap>;

function dijkstra(graph: Graph, start: string, end: string): [string[], number] {
  const dijkstra_table = Object.keys(graph).reduce((obj, node) => {
    obj[node] = Infinity;
    return obj;
  }, {} as EdgeMap);
  const record_path = Object.keys(graph).reduce((obj, node) => {
    obj[node] = null;
    return obj;
  }, {} as Record<string, string | null>);

  const visited = new Set<string>();

  // init
  dijkstra_table[start] = 0;
  const queue: [string, number][] = [[start, 0]];

  while (queue.length) {
    const [current_node, current_cost] = queue.shift()!;

    if (visited.has(current_node)) {
      continue;
    }
    visited.add(current_node);

    for (const neighbor in graph[current_node]) {
      const weight = graph[current_node][neighbor];
      if (current_cost + weight < dijkstra_table[neighbor]) {
        dijkstra_table[neighbor] = current_cost + weight;
        record_path[neighbor] = current_node;
        queue.push([neighbor, current_cost + weight]);
      }
    }
  }

  const getShortestPath = (node: string | null, result: string[]) => {
    if (!node) {
      return;
    }
    result.unshift(node);
    getShortestPath(record_path[node], result);
  };

  const shortest_path: string[] = [];
  getShortestPath(end, shortest_path);

  return [shortest_path, dijkstra_table[end]];
}

// test
const graph = {
  A: { B: 1, C: 1, D: 3 },
  B: { A: 1, D: 2, E: 1 },
  C: { A: 1, D: 1 },
  D: { A: 2, B: 3, C: 1, E: 2 },
  E: { B: 1, D: 2 },
};
const [result, cost] = dijkstra(graph, 'A', 'E');
console.log(result, cost);

File: Other/Review/Phase 6/Review2/Part1/dijkstra/dijkstra.go
package main

import (
	"fmt"
	"math"
	"sort"
)

// EdgeMap represents the neighbors and their corresponding weights
type EdgeMap map[string]int

// Graph represents the whole graph where each node maps to its neighbors
type Graph map[string]EdgeMap

// queue's element
type Item struct {
	node string
	cost int
}

func dijkstra(graph Graph, start string, end string) ([]string, int) {
	dijkstraTable := make(EdgeMap)
	recordPath := make(map[string]string)

	for key := range graph {
		dijkstraTable[key] = math.MaxInt
		recordPath[key] = ""
	}

	// init
	queue := make([]*Item, 0)
	visited := make(map[string]bool, 0)

	queue = append(queue, &Item{start, 0})
	dijkstraTable[start] = 0

	for len(queue) > 0 {
		sort.Slice(queue, func(i, j int) bool {
			return queue[i].cost < queue[j].cost
		})

		currentNode, currentCost := queue[0].node, queue[0].cost
		queue = queue[1:]

		if _, exists := visited[currentNode]; exists {
			continue
		}
		visited[currentNode] = true

		neighbors := graph[currentNode]
		for neighbor, weight := range neighbors {
			if currentCost+weight < dijkstraTable[neighbor] {
				dijkstraTable[neighbor] = currentCost + weight
				recordPath[neighbor] = currentNode
				queue = append(queue, &Item{neighbor, currentCost + weight})
			}
		}
	}

	shortestPath := []string{}
	var getShortestPath func(node string)
	getShortestPath = func(node string) {
		if node == "" {
			return
		}
		shortestPath = append(shortestPath, node)
		getShortestPath(recordPath[node])
	}

	getShortestPath(end)

	for i, j := 0, len(shortestPath)-1; i < j; i, j = i+1, j-1 {
		shortestPath[i], shortestPath[j] = shortestPath[j], shortestPath[i]
	}

	return shortestPath, dijkstraTable[end]
}

// test
func main() {
	graph := Graph{
		"A": {"B": 1, "C": 1, "D": 3},
		"B": {"A": 1, "D": 2, "E": 1},
		"C": {"A": 1, "D": 1},
		"D": {"A": 3, "B": 2, "C": 1, "E": 2},
		"E": {"B": 1, "D": 2},
	}

	result, cost := dijkstra(graph, "A", "E")
	fmt.Println("Shortest path:", result, "Cost:", cost)
}
